### PR TITLE
Fix softmax window masking

### PR DESCRIPTION
**Pull request by Spotlight0xff**
Hi,

this PR adds a test for the `SoftmaxOverSpatial` layer in the case of window masking.
Previously this functionality was broken due to some previous changes in the "default" case.
Some of the changes were already in some branch, but got lost I think.

Also, see
```python
indices.set_shape((None, None))  # required for older TF versions
```
which is a workaround for an error I got for TF 1.4.1, where the tensor shape could not get inferred automatically (and thus `nd_indices` failed). Hope this is ok.

The test is also good to verify various edge cases for the masking.

Cheers

### CODE REVIEW

File: TFNetworkLayer.py [hunk 1]
```diff
@@ -2867,14 +2868,24 @@ def __init__(self, axis=None, energy_factor=None,
...
-        mask_shape[axis] = window_len
-        energy_mask_window = tf.scatter_nd(idxs, tf.ones(shape=mask_shape), energy_shape[:2])
+        indices = expand_dims_unbroadcast(tf.range(window_len), 0, n_batch)  # (B, W)
+        # (B, W) + (B, 1)
+        indices += tf.expand_dims(tf.to_int32(window_start), axis=1)  # + (B, 1)
+        indices.set_shape((None, None))  # required for older TF versions
+        idxs = nd_indices(indices)  # (B, W, 2)
+        mask_shape = [n_batch, window_len]  # (B, W)
+        rem_energy_data = energy_data.copy_with_time_dim_axis(time_dim_axis=1)  # (B, T, ...)
+        rem_energy_shape = tf.shape(rem_energy_data.placeholder, name="rem_energy_shape")
```
Comment (albertz) [inline TFNetworkLayer.py]: Note that I prefer to use `TFUtil.get_shape` instead of `tf.shape` in many cases now.

File: TFNetworkLayer.py [hunk 2]
```diff
@@ -2867,14 +2868,24 @@ def __init__(self, axis=None, energy_factor=None,
...
+        rem_energy_shape = [rem_energy_shape[i] for i in range(rem_energy_data.batch_ndim)]
+        rem_axes = list(energy_shape)
+        rem_axes.pop(energy_data.batch_dim_axis)
+        rem_axes.pop()  # (B, ..., T) -> (...)
```
Comment (albertz) [inline TFNetworkLayer.py]: `pop` removes by default the first or last? Maybe better make that explicit.
Is the comment correct? Above you do `rem_energy_data. copy_with_time_dim_axis(1)`, so time is at axis 1, not at the end?
Also, does this code assume that batch dim axis is 0 or 1? What if it is bigger?

File: TFNetworkLayer.py (hunk 1 above)
Comment (Spotlight0xff) [reply TFNetworkLayer.py]: Cool, I changed that.

File: TFNetworkLayer.py (hunk 2 above)
Comment (Spotlight0xff) [reply TFNetworkLayer.py]: `pop` removes the last item by default. I made it more explicit.
`rem_axes.pop()` operates on a copy of `energy_shape`, and not `rem_energy_data`.
I've moved the code a bit further down, so that it is not confusing at first glance.

I haven't tested it with different batch-dim axes, but it does convert it into batch-major.

**Comment by Spotlight0xff** [comment]
Thanks, I addressed your comments, you can merge if it still looks good.
